### ISSUE TITLE
Disable private-member-access in the tests folder

### ISSUE THREAD

**Issue by AndreuCodina** [astral-sh/ruff#16892]
### Summary

We can't do good tests if we use this rule of Ruff. Let's put an example:

In this repository https://github.com/AndreuCodina/test-private-member I have the `UserService` class I want to test. An example of a test is to try to sign up an user, but we need to mock the private method `_associate_bank_account` because it accesses the bank and it's not something we want to test. So, we have this test:

```python
from pytest_mock import MockerFixture

from src.services.user_service import UserService


class TestUserService:
    def test_associate_bank_account(self, mocker: MockerFixture) -> None:
        user_service = UserService()

        associate_bank_account_mock = mocker.patch(
            f"{UserService.__module__}.{UserService._associate_bank_account.__qualname__}",
        )

        user_service.sign_up("fran@gmail.com")

        associate_bank_account_mock.assert_called_once_with()
```

Developers usually use

```python
associate_bank_account_mock = mocker.patch(
    "src.services.user_service.UserService._associate_bank_account"
)
```

but it's a really bad practice: they're using magic strings. So, for this reason I'm using `__module__` and `__qualname__`, but Ruff complains with this rule: https://docs.astral.sh/ruff/rules/private-member-access/

This rule should be disabled when we are in the `tests` folder.

**Comment by MichaReiser** [comment]
I can see how this is frustrating, but I'm not entirely sure whether disabling the rule here is necessarily the right solution because it also silences all private member access for types for which you may want to detect the access still. There's also the argument that the accessors for mocking should be public (if that's the intended API). I don't want to discuss whether that's best practice, but I could see people at either end of that argument.

That's why I'm somewhat leaning towards users making an explicit call on whether they want to allow this rule for their test code (use `per-file-ignore`) or not.

**Comment by AndreuCodina** [comment]
> I can see how this is frustrating, but I'm not entirely sure whether disabling the rule here is necessarily the right solution because it also silences all private member access for types for which you may want to detect the access still. There's also the argument that the accessors for mocking should be public (if that's the intended API). I don't want to discuss whether that's best practice, but I could see people at either end of that argument.
> 
> That's why I'm somewhat leaning towards users making an explicit call on whether they want to allow this rule for their test code (use `per-file-ignore`) or not.

Could you provide an example of a case where you want to detect the access from the `tests` folder? I can only think about preventing access from the `src` folder, not `tests`.
I'm speaking about the `src` folder of the repository, not external libraries, in case you were thinking of that case.

For example, in .NET, we edit the pyproject.toml equivalent to say "the private code of each folder will be visible to its test folder", but IMO it should be the default (this ecosystem usually doesn't make breaking changes).

I don't see that adding an ignore in each file is better than making all methods public.

**Comment by MichaReiser** [comment]
> Could you provide an example of a case where you want to detect the access from the tests folder? I can only think about preventing access from the src folder, not tests.

I'd be more hesitant to allow private access to modules that come from third-party dependencies.

> I don't see that adding an ignore in each file is better than making all methods public.

You can ignore the rule at the project level using [`per-file-ignores`](https://docs.astral.sh/ruff/settings/#lint_per-file-ignores)

**Comment by AndreuCodina** [comment]
Cool! I'll tell my company, and let them choose.

But, in my opinion, this is technical debt. In my company, one of the top complaints of Python projects is the amount of configurations to create a new project, and also update them later.
